File: dart/src/linkedList/CircularLinkedList/CircularLinkedList.dart
class CircularLinkedListNode<T> {
  T data;
  CircularLinkedListNode<T>? next;
  CircularLinkedListNode<T>? prev;

  CircularLinkedListNode(this.data) {}
}

class CircularLinkedList<T> {
  CircularLinkedListNode<T>? head;
  int? length;

  CircularLinkedList() {
    this.head = null;
    this.length = 0;
  }

  get size => this.length;
  get isEmpty => this.length == 0;
  isCircular() {
    var flag = false;
    var current = this.head;
    while (current != null) {
      if (current.next == this.head) {
        flag = true;
        break;
      }
      current = current.next;
    }
    return flag;
  }

  void append(T data) {
    final node = CircularLinkedListNode<T>(data);

    if (this.head == null) {
      this.head = node;
      node.next = node;
      node.prev = node;
    } else {
      var tail = this.head?.prev;
      tail?.next = node;
      node.prev = tail;
      node.next = this.head;
      this.head?.prev = node;
    }
    this.length = this.length! + 1;
  }

  void prepend(T data) {
    final node = CircularLinkedListNode<T>(data);

    if (this.head == null) {
      this.head = node;
      node.next = node;
      node.prev = node;
    } else {
      node.next = this.head;
      node.prev = this.head?.prev;
      this.head?.prev?.next = node;
      this.head?.prev = node;
      this.head = node;
    }
    this.length = this.length! + 1;
  }

  void insertAt(T data, int index) {
    if (index < 0 || index > this.length!) {
      throw RangeError("Invalid index");
    }
    final node = CircularLinkedListNode<T>(data);

    if (index == 0) {
      if (head == null) {
        node.next = node;
        node.prev = node;
      } else {
        node.next = this.head;
        node.prev = this.head?.prev;
        this.head?.prev?.next = node;
        this.head?.prev = node;
      }
      this.head = node;
    } else {
      var current = this.head;
      for (int i = 0; i < index - 1; i++) {
        current = current?.next;
      }
      node.next = current?.next;
      node.prev = current;
      current?.next?.prev = node;
      current?.next = node;
    }
    this.length = this.length! + 1;
  }

  void deleteAt(int index) {
    if (index < 0 || index >= this.length!) {
      throw RangeError("Invalid index");
    }

    if (index == 0) {
      if (this.length == 1) {
        this.head = null;
      } else {
        var next = this.head?.next;
        next?.prev = this.head?.prev;
        this.head?.prev?.next = next;
        this.head = next;
      }
    } else {
      var current = this.head;
      for (int i = 0; i < index - 1; i++) {
        current = current?.next;
      }
      var next = current?.next?.next;
      next?.prev = current;
      current?.next = next;
    }
    this.length = this.length! - 1;
  }

  toList() {
    List<T> list = [];
    CircularLinkedListNode<T>? current = this.head;
    while (current != null) {
      list.add(current.data);
      current = current.next;
      if (current == this.head) {
        break;
      }
    }
    return list;
  }

  reverseToList() {
    List<T> list = [];
    CircularLinkedListNode<T>? current = this.head?.prev;
    while (current != null) {
      list.add(current.data);
      current = current.prev;
      if (current == this.head?.prev) {
        break;
      }
    }
    return list;
  }
}

File: go/linkedlist/CircularLinkedList/CircularLinkedList.go
package CircularLinkedList

import "fmt"

type CircularLinkedListNode struct {
	Next  *CircularLinkedListNode
	Value int
	Prev  *CircularLinkedListNode
}

type CircularLinkedList struct {
	Head *CircularLinkedListNode

	Length int
}

func NewCircularLinkedList() *CircularLinkedList {
	return &CircularLinkedList{
		Head:   nil,
		Length: 0,
	}
}

func (l *CircularLinkedList) Prepend(value int) {
	node := &CircularLinkedListNode{
		Value: value,
		Next:  nil,
		Prev:  nil,
	}

	if l.Head == nil {
		l.Head = node
		node.Next = node
		node.Prev = node
	} else {
		node.Next = l.Head
		node.Prev = l.Head.Prev
		l.Head.Prev.Next = node
		l.Head.Prev = node
		l.Head = node
	}
	l.Length++
}
func (l *CircularLinkedList) IsEmpty() bool {
	return l.Length == 0
}
func (l *CircularLinkedList) IsCircular() bool {
	if l.Head == nil {
		return false
	}
	temp := l.Head
	for temp != nil {
		if temp.Next == l.Head {
			return true
		}
		temp = temp.Next
	}
	return false
}

func (l *CircularLinkedList) DeleteAt(index int) (err error) {
	if index < 0 || index >= l.Length {
		return fmt.Errorf("index out of range")
	}
	if index == 0 {
		if l.Length == 1 {
			l.Head = nil
		} else {
			temp := l.Head
			l.Head = l.Head.Next
			l.Head.Prev = temp.Prev
			temp.Prev.Next = l.Head
		}
	} else {
		temp := l.Head
		for i := 0; i < index-1; i++ {
			temp = temp.Next
		}
		temp.Next = temp.Next.Next
		temp.Next.Prev = temp
	}
	l.Length--
	return nil
}

func (l *CircularLinkedList) Append(value int) {
	node := &CircularLinkedListNode{
		Value: value,
		Next:  nil,
		Prev:  nil,
	}

	if l.Head == nil {
		l.Head = node
		node.Next = node
		node.Prev = node
	} else {
		tail := l.Head.Prev
		node.Next = l.Head
		node.Prev = tail
		tail.Next = node
		l.Head.Prev = node
	}
	l.Length++
}

func (l *CircularLinkedList) InsertAt(value int, index int) (err error) {
	if index < 0 || index > l.Length {
		return fmt.Errorf("index out of range")
	}
	node := &CircularLinkedListNode{
		Value: value,
		Next:  nil,
		Prev:  nil,
	}
	if index == 0 {
		if l.Head == nil {
			l.Head = node
			node.Next = node
			node.Prev = node
		} else {
			temp := l.Head
			node.Next = temp
			node.Prev = temp.Prev
			temp.Prev.Next = node
			temp.Prev = node
			l.Head = node
		}

	} else {
		temp := l.Head
		for i := 0; i < index-1; i++ {
			temp = temp.Next
		}
		node.Next = temp.Next
		node.Prev = temp
		temp.Next.Prev = node
		temp.Next = node
	}
	l.Length++
	return nil
}

func (l *CircularLinkedList) ToArray() []int {
	if l.Head == nil {
		return []int{}
	}
	temp := l.Head
	arr := make([]int, l.Length)
	for i := 0; i < l.Length; i++ {
		arr[i] = temp.Value
		temp = temp.Next
		if temp == l.Head {
			break
		}
	}
	return arr
}

func (l *CircularLinkedList) ReverseToArray() []int {
	if l.Head == nil {
		return []int{}
	}
	temp := l.Head.Prev
	arr := make([]int, l.Length)
	for i := 0; i < l.Length; i++ {
		arr[i] = temp.Value
		temp = temp.Prev
		if temp == l.Head.Prev {
			break
		}
	}
	return arr
}
